arch/arc/core/mpu/arc_mpu.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2017 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <soc.h>
+#include <arch/arc/v2/aux_regs.h>
+#include <arch/arc/v2/mpu/arc_mpu.h>
+#include <arch/arc/v2/mpu/arc_core_mpu.h>
+#include <arch/arc/v2/ffs.h>
+#include <logging/sys_log.h>
+
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
+#define AUX_MPU_EN_ENABLE   (0x40000000)
+#define AUX_MPU_EN_DISABLE  (0x0)
+
+#define AUX_MPU_RDP_REGION_SIZE(bits)  \
+			(((bits - 1) & 0x3) | (((bits - 1) & 0x1C) << 7))
+
+#define _ARC_V2_MPU_EN   (0x409)
+#define _ARC_V2_MPU_RDB0 (0x422)
+#define _ARC_V2_MPU_RDP0 (0x423)
+
+/* For MPU version 2, the minimum protection region size is 2048 bytes */
+/* FOr MPU version 3, the minimum protection region size is 32 bytes */
+#if CONFIG_ARC_MPU_VER == 2
+#define ARC_FEATURE_MPU_ALIGNMENT_BITS 11
+#elif CONFIG_ARC_MPU_VER == 3
+#define ARC_FEATURE_MPU_ALIGNMENT_BITS 5
+#endif
+
+
+static inline u8_t _get_num_regions(void)
+{
+	u32_t type = _arc_v2_aux_reg_read(_ARC_V2_MPU_BUILD);
+
+	type = (type & 0xFF00) >> 8;
+
+	return (u8_t)type;
+}
+
+/**
+ * @brief get the bits according to size
+ * @param[in] size
+ * @return bits
+ */
+static u32_t _vmpu_region_bits(u32_t size)
+{
+	u32_t bits;
+
+	bits = find_msb_set(size);
+
+	/* minimum region size is 2048 bytes */
+	if (bits < ARC_FEATURE_MPU_ALIGNMENT_BITS) {
+		bits = ARC_FEATURE_MPU_ALIGNMENT_BITS;
+	}
+
+	/* round up if needed */
+	if ((1UL << bits) < size) {
+		bits++;
+	}
+
+	return bits;
+}
+
+
+/**
+ * This internal function is utilized by the MPU driver to parse the intent
+ * type (i.e. THREAD_STACK_REGION) and return the correct parameter set.
+ */
+static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
+{
+	switch (type) {
+	case THREAD_STACK_REGION:
+		return 0;
+	case THREAD_STACK_GUARD_REGION:
+	/* no RW to guard region */
+		return  AUX_MPU_RDP_REGION_SIZE(
```

Comment:
should be
return AUX_MPU_RDP_REGION_SIZE(ARC_FEATURE_MPU_ALIGNMENT_BITS);
or
return AUX_MPU_RDP_REGION_SIZE(_vmpu_region_bits(1 << ARC_FEATURE_MPU_ALIGNMENT_BITS));
?